Add @SubclassOptInRequired documentation (#4414)
JetBrains/kotlin-web-site: docs/topics/opt-in-requirements.md

annotation class MyDateTime

@MyDateTime                            
class DateProvider // A class requiring opt-in
```

In your code, before declaring a function that uses the `DateProvider` class, add the `@OptIn` annotation with
a reference to the `MyDateTime` annotation class:

```kotlin
// Client code
@OptIn(MyDateTime::class)
fun getDate(): Date { // Uses DateProvider
    val dateProvider: DateProvider
    // ...
}
```

It's important to note that with this approach, if the `getDate()` function is called elsewhere in your code or used by 
another developer, no opt-in is required:

```kotlin
// Client code
@OptIn(MyDateTime::class)
fun getDate(): Date { // Uses DateProvider
    val dateProvider: DateProvider
    // ...
}

fun displayDate() {
    println(getDate()) // OK: No opt-in is required
}
```

The opt-in requirement is not propagated in your code. This is risky because you and others could use experimental API 
without realizing it. In the worst case, you could end up building a supposedly stable API on top of an experimental one.
A safer approach is to propagate the opt-in requirements when you choose to opt in.

> #### Propagate opt-in requirements

Author: Added!

When you use API in your code that's intended for third-party use, such as in a library, you can propagate its opt-in requirement

> to your API as well. To do this, annotate your declaration with the _[opt-in requirement annotation](#create-opt-in-requirement-annotations)_ of the API used in its body.

Author: Would this work as an alternative?

> To do this, mark your declaration with the same _[opt-in requirement annotation](#create-opt-in-requirement-annotations)_ used by the library.

class DateProvider // A class requiring opt-in
```

In your code, before declaring a function that uses the `DateProvider` class, you must add the `@MyDateTime` annotation:

```kotlin
// Client code
@MyDateTime
fun getDate(): Date {  
    val dateProvider: DateProvider // OK: the function requires opt-in as well
    // ...
}

fun displayDate() {
    println(getDate()) // Error: getDate() requires opt-in
}
```

As you can see in this example, the annotated function appears to be a part of the `@MyDateTime` API.
The opt-in propagates the opt-in requirement to users of the `getDate()` function.

> Note, if you implicitly use an API element that has an opt-in requirement, you are still required to opt-in. For example,

Author: Could you help me with a better example for implicitly using an API element?

if you use an API element that doesn't have an opt-in requirement annotation but its signature includes a type that requires
opt-in, using it triggers an error.

```kotlin
// Client code
fun getDate(dateProvider: DateProvider): Date { // Error: DateProvider requires opt-in
    // ...
}

fun displayDate() {

>     println(getDate()) // Warning: the signature of getDate() contains DateProvider, which requires opt-in

Author: Can you share a version of the updated code sample with that line? I'm not sure how to add the new `getDate()` function that you suggest.

}
```

To use multiple APIs that require opt-in, mark the declaration with all their opt-in requirement annotations. For example:

```kotlin

> @MyDateTime

Author: Could you please suggest one that I can add here?

 level = RequiresOptIn.Level.WARNING,
 message = "Interfaces in this library are experimental"
)
annotation class UnstableApi()

@SubclassOptInRequired(UnstableApi::class)
interface CoreLibraryApi // An interface requiring opt-in to extend
```

In your code, before creating a new interface that inherits from the `CoreLibraryApi` interface, add the `@SubclassOptInRequired`
annotation with a reference to the `UnstableApi` annotation class:

```kotlin
// Client code
@SubclassOptInRequired(UnstableApi::class)
interface SomeImplementation : CoreLibraryApi
```

Note that when you use the `@SubclassOptInRequired` annotation on a class, the opt-in requirement is not propagated to 
any [inner or nested classes](nested-classes.md):

```kotlin
// Library code
@RequiresOptIn
annotation class ExperimentalFeature

@SubclassOptInRequired(ExperimentalFeature::class)
open class FileSystem {
    open class File
}

// Client code
class NetworkFileSystem : FileSystem() // Opt-in is required

// Nested class
class TextFile : FileSystem.File() // No opt-in required
```

Alternatively, you can also opt in by using the `@OptIn` annotation, or an annotation that references the opt-in annotation

> class to propagate the requirement further in your code:

Author: Does this sound better?
> Alternatively, you can opt in by using the @OptIn annotation. You can also use an annotation that references the opt-in annotation class to propagate the requirement further to any uses of the class in your code.

To opt in to multiple APIs on the module level, add one of the described arguments for each opt-in requirement marker used in your module.

> ## Require opt-in for API 

Reviewer: could we have a short description similar to the one in Opt in to API﻿ section?

Author: Do you mean something like:

```
If a library author marks a declaration from their library's API as requiring opt-in to extend, you must give explicit consent before you can extend it in your code.
```

Reviewer: exactly! 👍

requires an opt-in:

```kotlin
// Library code
@RequiresOptIn(
 level = RequiresOptIn.Level.WARNING,
 message = "Interfaces in this library are experimental"
)
annotation class UnstableApi()

@SubclassOptInRequired(UnstableApi::class)
// An interface requiring opt-in to extend
interface CoreLibraryApi 
```

In your code, before creating a new interface that inherits from the `CoreLibraryApi` interface, add the `@SubclassOptInRequired`
annotation with a reference to the `UnstableApi` annotation class:

```kotlin
// Client code
@SubclassOptInRequired(UnstableApi::class)
interface SomeImplementation : CoreLibraryApi
```

Note that when you use the `@SubclassOptInRequired` annotation on a class, the opt-in requirement is not propagated to 
any [inner or nested classes](nested-classes.md):

```kotlin
// Library code
@RequiresOptIn
annotation class ExperimentalFeature

@SubclassOptInRequired(ExperimentalFeature::class)
open class FileSystem {
    open class File
}

// Client code

// Opt-in is required
class NetworkFileSystem : FileSystem()

// Nested class
// No opt-in required
class TextFile : FileSystem.File()
```

> Alternatively, you can opt in by using the `@OptIn` annotation. You can also use an experimental marker annotation itself

Reviewer: ```suggestion
Alternatively, you can opt in by using the `@OptIn` annotation. You can also use an experimental marker annotation
```

annotation class MyDateTime
```

Opt-in requirement annotations must meet several requirements. They must have:

> * `BINARY` or `RUNTIME` [retention](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.annotation/-retention/).

Reviewer: (style) 
To align with other ULs could we have an enter here?

* `EXPRESSION`, `FILE`, `TYPE`, or `TYPE_PARAMETER` as a [target](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.annotation/-target/).
* No parameters.

An opt-in requirement can have one of two severity [levels](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-requires-opt-in/-level/):

> * `RequiresOptIn.Level.ERROR`. Opt-in is mandatory. Otherwise, the code that uses marked API won't compile. This is the default level.

Reviewer: (style) 
(same here)To align with other ULs could we have an enter here?

```

Note that for some language elements, an opt-in requirement annotation is not applicable:

* You can't annotate a backing field or a getter of a property, just the property itself.
* You can't annotate a local variable or a value parameter.

## Require opt-in to extend API

There may be times when you want more granular control over which specific parts of your API can be used and
extended. For example, when you have some API that is stable to use but:

* **Unstable to implement** due to ongoing evolution, such as when you have a family of interfaces where you expect to add new abstract functions without default implementations.

> * **Closed to third-party implementations** due to internal or technical reasons, such as an API that you want to be sealed but can't for technical reasons.

Author: Ok, removed it!

In such cases, you want to require users to opt in to your API before they can extend it further.
For example, by inheriting from the API or declaring implementations for abstract functions. Using the [`@SubclassOptInRequired`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-subclass-opt-in-required/)
annotation, you can enforce this behavior for [open](inheritance.md) or [abstract classes](classes.md#abstract-classes) and [non-functional interfaces](interfaces.md).

To add the opt-in requirement to an API element, use the [`@SubclassOptInRequired`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-subclass-opt-in-required/)
annotation with a reference to the annotation class:

```kotlin
@RequiresOptIn(
 level = RequiresOptIn.Level.WARNING,
 message = "Interfaces in this library are experimental"
)
annotation class UnstableApi()

@SubclassOptInRequired(UnstableApi::class)
interface CoreLibraryApi // An interface requiring opt-in to extend
```

> Note that when you use the `@SubclassOptInRequired` annotation to require opt-in, the requirement is not propagated to

Author: Yes. Since the doc is split into two parts - one for those who want to opt in to API and one for those who want to add opt-in requirements for their API. This way if only one part is read, the information isn't missed. It seems relevant for both use cases.

```

Note that for some language elements, an opt-in requirement annotation is not applicable:

* You can't annotate a backing field or a getter of a property, just the property itself.
* You can't annotate a local variable or a value parameter.

## Require opt-in to extend API

There may be times when you want more granular control over which specific parts of your API can be used and
extended. For example, when you have some API that is stable to use but:

* **Unstable to implement** due to ongoing evolution, such as when you have a family of interfaces where you expect to add new abstract functions without default implementations.
* **Delicate or fragile to implement**, such as individual functions that need to behave in a coordinated manner.

> * **Has a contract that may be weakened in the future** in a backwards-incompatible manner for external implementations, such as changing an input parameter `T` to a nullable version `T?` where the code didn't previously consider `null` values.

Reviewer: ```suggestion
* **Has a contract that may be weakened in the future** in a backward-incompatible way for external implementations, such as changing an input parameter `T` to a nullable version `T?` where `null` values were not previously considered.
```

> In such cases, you want to require users to opt in to your API before they can extend it further.

Reviewer: Hmm 🤔 
How about?
In such cases, you might want to require users
or 
In such cases, you can....

Author: I like "can".

> For example, by inheriting from the API or declaring implementations for abstract functions. Using the [`@SubclassOptInRequired`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-subclass-opt-in-required/)

Reviewer: Hmm 🤔 I'm a bit confused by this sentence: For example, by inheriting from the API or declaring implementations for abstract functions. 
what do you think about having this together with the previous sentence/ somehow rewording to connect it more? 
....extend it further. Extending the API includes, inheriting from the API or declaring implementations...

Author: What about:

> In such cases, you can require users to opt in to your API before they can extend it further. Users can extend your API by inheriting from the API or implementing abstract functions. By using the `@SubclassOptInRequired` annotation, you can enforce this requirement to opt-in for open or abstract classes and non-functional interfaces.

Reviewer: nice one! 👍  I like it!

To add the opt-in requirement to an API element, use the [`@SubclassOptInRequired`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-subclass-opt-in-required/)
annotation with a reference to the annotation class:

```kotlin
@RequiresOptIn(
 level = RequiresOptIn.Level.WARNING,
 message = "Interfaces in this library are experimental"
)
annotation class UnstableApi()

@SubclassOptInRequired(UnstableApi::class)
// An interface requiring opt-in to extend
interface CoreLibraryApi 
```

Note that when you use the `@SubclassOptInRequired` annotation to require opt-in, the requirement is not propagated to
any [inner or nested classes](nested-classes.md).

For a real-world example of how to use the `@SubclassOptInRequired` annotation in your API, check out the
[`SharedFlow`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-shared-flow/)
interface in the `kotlinx.coroutines` library.

## Opt-in requirements for pre-stable APIs

If you use opt-in requirements for features that are not stable yet, carefully handle the API graduation to avoid 
breaking client code.

> Once your pre-stable API graduates and is released in a stable state, remove any opt-in requirement annotations from 

Reviewer: ```suggestion
Once your pre-stable API graduates and is released in a stable state, remove opt-in requirement annotations from 
```

> your declarations. The clients can then use them without restriction. However, you should leave the annotation classes 

Reviewer: ```suggestion
your declarations. This allows clients to use them without restrictions. However, you should leave the annotation classes 
```

Reviewer: Do you think we should keep in "should" here?
I think it might be better to remove it to ensure compatibility.

Author: @BlondeHex do you mind if we remove "should" here? Or should we leave it like this because it's intended as a recommendation and not a hard rule?

in modules so that the existing client code remains compatible.

To encourage API users to update their modules by removing any annotations from their code and recompiling, mark the annotations
as [`@Deprecated`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-deprecated/) and provide an explanation in the deprecation message.

```kotlin

> @Deprecated("This opt-in requirement is not used anymore. Remove its usages from your code.")

Reviewer: Could you please add +1 enter to the bottom of the page?

Author: Sure. I'm curious though. Why? :)

Reviewer: I'm not sure why JB docs follow this convention, I've seen that most of our docs follow it.  Personally, my IDE always sends me an error if I don't put a new line at the end of a text file because of posix compliance
